### PR TITLE
DataSourcePicker: Add recently used from local storage to ds picker

### LINKED ISSUE #66944: Reimplement the sorting of data sources.
The new data source picker should sort the data sources so that the data sources most relevant to the user ends up near the top.

### PR DESCRIPTION
**Pull request by oscarkilhed**
**What is this feature?**
Adds local storage backed sorting of data sources by recently used.

Fixes #66944

### CODE REVIEW

File: public/app/features/datasources/components/picker/utils.ts [hunk 1]
```diff
@@ -40,9 +44,25 @@ export function dataSourceLabel(
...
+  const [value = [], setStorage] = useLocalStorage<string[]>(LOCAL_STORAGE_KEY, []);
+
   const dataSourceSrv = getDataSourceSrv();
 
-  return dataSourceSrv.getList(filters);
+  const dataSources = dataSourceSrv.getList(filters);
+
+  const updateStorage = (ds: DataSourceInstanceSettings) => {
+    if (ds.meta.builtIn) {
+      return; //Prevent storing the built in datasources (-- Grafana --, -- Mixed --,  -- Dashboard --)
```
Comment (baldm0mma) [inline public/app/features/datasources/components/picker/utils.ts:55]: Great comment!!!

**Review by baldm0mma — APPROVED** [review]
Genius work!!! ;)

File: public/app/features/datasources/components/picker/DataSourceList.tsx [hunk 2]
```diff
@@ -35,7 +35,7 @@ export interface DataSourceListProps {
 export function DataSourceList(props: DataSourceListProps) {
   const { className, current, onChange } = props;
   // QUESTION: Should we use data from the Redux store as admin DS view does?
-  const dataSources = useGetDatasources({
+  const { dataSources, updateStorage } = useGetDatasources({
```
Comment (ivanortegaalba) [inline public/app/features/datasources/components/picker/DataSourceList.tsx:38]: I'm not very convinced about the hook API 🤔 

Just renaming the function, could be already an improvement
```suggestion
  const { dataSources, pushRecentlyUsed } = useGetDatasources({
```

However, this hook is hard to use in other places outside of the picker. Maybe just splitting the hooks in two would leave a single responsibility to each of them:
```js
const { dataSources } = useDatasources(filters)
const { recentlyUsed, pushRecentlyUsedItem } = useRecentlyUsedDatasources();
```
Then, you compose as you need them. No strong opinion. Thoughts @oscarkilhed?

Comment (ivanortegaalba) [reply public/app/features/datasources/components/picker/DataSourceList.tsx:38]: Also, thinking in requirements like "move to the top the currently selected element". Probably it makes sense to leave the sorting to the component displaying the list. WDYT?

File: public/app/features/datasources/components/picker/DataSourceList.tsx [hunk 3]
```diff
@@ -48,18 +48,33 @@ export function DataSourceList(props: DataSourceListProps) {
     variables: props.variables,
   });
 
+  const { recentlyUsedDataSources, pushRecentlyUsedDataSource } = useRecentlyUsedDataSources();
+
+  const orderedDataSources = orderDataSourcesByRecentlyUsed(recentlyUsedDataSources, dataSources);
```
Comment (ivanortegaalba) [inline public/app/features/datasources/components/picker/DataSourceList.tsx:53]: Here, we should also move the current selected DS to the top. 

The order should be:
- Current
- Recently used
- Variables
- Rest is sorted alphabetically.

So I'd rename this to `sortDataSources(...)` and delegate everything to that function. WDYT?

File: public/app/features/datasources/components/picker/utils.ts [hunk 4]
```diff
@@ -39,13 +43,42 @@ export function dataSourceLabel(
...
+        value.findIndex((dsUid) => ds.uid === dsUid),
+        1
+      );
+      setStorage([ds.uid, ...value]);
+    } else {
+      setStorage([ds.uid, ...value].slice(0, 5));
+    }
+  };
+
+  return { recentlyUsedDataSources: value, pushRecentlyUsedDataSource };
```
Comment (ivanortegaalba) [inline public/app/features/datasources/components/picker/utils.ts:71]: Should we return them as array, as other hooks do? This allows renaming the props depending on the context.

File: public/app/features/datasources/components/picker/DataSourceList.tsx (hunk 3 above)
Comment (oscarkilhed) [reply public/app/features/datasources/components/picker/DataSourceList.tsx:53]: Sounds good. I'll implement it all in that function.

**Comment by ivanortegaalba** [comment]
Also, this is a good scenario to write some test and ensure the hooks works as expected. They could be under `app/features/datasources/hooks`

File: public/app/features/datasources/components/picker/utils.test.ts [hunk 5]
```diff
@@ -28,3 +28,47 @@ describe('isDataSourceMatch', () => {
     expect(isDataSourceMatch(dataSourceInstanceSettings, { uid: 'b' } as DataSourceRef)).toBeFalsy();
   });
 });
+
+describe('getDataSouceCompareFn', () => {
```
Comment (ivanortegaalba) [inline public/app/features/datasources/components/picker/utils.test.ts:32]: Maybe, I'd add a use case to make sure all of them work together. Something like: 
```js
  it('should sort alphabetically ordering the groups like: current > Variables > Rest > Built-in', ...);
```

So we can guarantee that there are two variables or that the built-in DS is also sorted properly. But also that the groups are placed in order as well.

In summary, I think we can simplify the unit tests into one 😄

**Review by ivanortegaalba — APPROVED** [review]
Great work Oscar! 💯

**Pull request merged**